docs: converter (#1)
sdsc-ordes/catplus-docs: docs/converters.md

# Converter

The Cat+ converter is at this repository [catplus-converters](https://github.com/sdsc-ordes/catplus-converters).

## Architecture

> 

Reviewer: ```suggestion
The repository contains 3 crates under `src`:
* catplus-common: primitives and mappings for the swisscatplus data model.
* converter: command line interface for the converter.
* validation: command line helper for SHACL validation of RDF graphs.
```

> ### Common lib Crate

Reviewer: ```suggestion
### catplus-commonf
```

> The common lib is where all the mapping happens, from the raw JSON objects to the RDF terms.

Reviewer: ```suggestion
The catplus-common crate is where all the mapping happens, from the raw JSON objects to the RDF terms.
```

> The RDF parsing and serialization for both turtle and jsonld formats happen in the scripts of the `rdf` folder. 

Reviewer: ```suggestion
The RDF parsing and serialization for both turtle and jsonld formats happen in the scripts of the `rdf` module. 
```

> The `models` folder splits the representation of instance data by file type (with list-like concepts such as units under `enums` and overarching terms repeating over file types in `core`). Objects are defined as structs, with each one implementing the trait `InsertIntoGraph`. If a unique URI is required for the object (as for devices, chemicals, products), then the function `get_uri` must be defined within the trait (see existing examples in Agilent and Synth). The [data linkage documentation](data-linkage.md) can help understand how the different models come together.

Reviewer: ```suggestion
The `models` module splits the representation of instance data by file type (with list-like concepts such as units under `enums` and overarching terms repeating over file types in `core`). Objects are defined as structs, with each one implementing the trait `InsertIntoGraph`. If a unique URI is required for the object (as for devices, chemicals, products), then the function `get_uri` must be defined within the trait (see existing examples in Agilent and Synth). The [data linkage documentation](data-linkage.md) can help understand how the different models come together.
```

> The `graph` folder contains the `namespaces` folder where each prefix and its objects are defined. In the same folder, the `insert_into` script defines how different types of objects get inserted into the graph to produce valid RDF. The `graph_builder` script defines the way a graph is built by calling the insert into functions. It is also the place where the `contentURL` gets inserted into the graph. The `contentURL` is the path to the raw data file. Currently, the raw data is linked to the metadata at the level of the HCI file, onto the Campaign object, as well as at the Agilent level, onto the Liquid Chromatography Document object. New contentURL objects can be defined in the `link_content` function following the existing examples. The raw data file path of the file being parsing will be automatically attached under this concept: hence this concept needs to be unique to the file being parsed for their to be a unique link to the raw data.

Reviewer: ```suggestion
The `graph` module contains the `graph::namespaces` child module where each prefix and its objects are defined. The `graph::insert_into` module defines how different types of objects get inserted into the graph to produce valid RDF. The `graph::graph_builder` exposes a Builder interface wrapping the insert_into methods. It is also where the `schema:contentURL` field gets inserted into the graph. The `schema:contentURL` is the path to the raw data file. Currently, the raw data is linked to the metadata at the level of the HCI file, onto the Campaign object, as well as at the Agilent level, onto the Liquid Chromatography Document object. New contentURL objects can be defined in the `link_content` function following the existing examples. The raw data file path of the file being parsing will be automatically attached under this concept: hence this concept needs to be unique to the file being parsed for their to be a unique link to the raw data.
```

### Converter Crate

The converter can process file one by one or an entire folder.

Example for a file:

```just
just convert /examples/2-Agilent.json turtle
```

Example for a folder (multifiles):

```just
just convert /examples/complex_model/Bravo2/ turtle
```

Example for a folder (multifiles) where the output folder is specified:

```just
just convert /examples/complex_model/Bravo2/ turtle --output-folder=examples/rdf
```

`main` defines the CLI processing for the converter. The converter infers automatically from the raw data filename which type of conversion it needs to perform.

> `converter` takes care of instanciating a graph, building the graph by parsing the JSON, and finally linking in the raw data file path if relevant (under `contentURL`).

Reviewer: ```suggestion
`converter` takes care of instantiating a graph, building the graph by parsing the JSON, and finally linking in the raw data file path if relevant (under `contentURL`).
```

fix: typo

# Converter

The Cat+ converter is at this repository [catplus-converters](https://github.com/sdsc-ordes/catplus-converters).

## Architecture

The repository contains 3 crates under `src`:

* catplus-common: primitives and mappings for the swisscatplus data model.
* converter: command line interface for the converter.
* validation: command line helper for SHACL validation of RDF graphs.

### catplus-common

The catplus-common crate is where all the mapping happens, from the raw JSON objects to the RDF terms.

The RDF parsing and serialization for both turtle and jsonld formats happen in the scripts of the `rdf` module. 

> The `models` module splits the representation of instance data by file type (with list-like concepts such as units under `enums` and overarching terms repeating over file types in `core`). Objects are defined as structs, with each one implementing the trait `InsertIntoGraph`. If a unique URI is required for the object (as for devices, chemicals, products), then the function `get_uri` must be defined within the trait (see existing examples in Agilent and Synth). The [data linkage documentation](data-linkage.md) can help understand how the different models come together.

Reviewer 2: ```suggestion
In `models` the json input data is mapped with `serde-json` to rust structs. The rust structs then construct the rdf graph by each implementing a trait `InsertIntoGraph`. There are several json input files, see  [data linkage documentation](data-linkage.md). Accordingly `models` has several submodules:

* `core`: contains the transformation of structs that are shared between multiple files
* `enum`: contains the mapping of enumeration types, especially the mapping of action types as most json input files contain actions, that are mapped to different action classes in the ontology.
* `agilent`, `synth`, `bravo`, `hci`: contain transformations that are specific for the corresponding input data files

If a unique URI is required for the object (as for devices, chemicals, products), then the function `get_uri` must be defined within the trait (see existing examples in Agilent and Synth). The [data linkage documentation](data-linkage.md) can help understand how the different models come together.
```

> The `graph` module contains the `graph::namespaces` child module where each prefix and its objects are defined. The `graph::insert_into` module defines how different types of objects get inserted into the graph to produce valid RDF. The `graph::graph_builder` exposes a Builder interface wrapping the insert_into methods. It is also where the `schema:contentURL` field gets inserted into the graph. The `schema:contentURL` is the path to the raw data file. Currently, the raw data is linked to the metadata at the level of the HCI file, onto the Campaign object, as well as at the Agilent level, onto the Liquid Chromatography Document object. New contentURL objects can be defined in the `link_content` function following the existing examples. The raw data file path of the file being parsing will be automatically attached under this concept: hence this concept needs to be unique to the file being parsed for their to be a unique link to the raw data.

Reviewer 2: ```suggestion
The `graph` module contains the `graph::namespaces` child module where each prefix and its terms are defined. The `graph::insert_into` module defines how different types of objects get inserted into the graph to produce valid RDF. The `graph::graph_builder` exposes a Builder interface wrapping the insert_into methods. It is also where the `schema:contentURL` field gets inserted into the graph. The `schema:contentURL` is the path to the raw data json input file. Currently, the raw data is linked to the metadata at the level of the HCI file, onto the Campaign object, as well as at the Agilent level, onto the Liquid Chromatography Document object. New contentURL objects can be defined in the `link_content` function following the existing examples. The raw data file path of the file being parsing will be automatically attached under this concept: hence this concept needs to be unique to the file being parsed for their to be a unique link to the raw data.
```

### Converter Crate

The converter can process file one by one or an entire folder.

Example for a file:

```just
just convert /examples/2-Agilent.json turtle
```

Example for a folder (multifiles):

```just
just convert /examples/complex_model/Bravo2/ turtle
```

Example for a folder (multifiles) where the output folder is specified:

```just
just convert /examples/complex_model/Bravo2/ turtle --output-folder=examples/rdf
```

`main` defines the CLI processing for the converter. The converter infers automatically from the raw data filename which type of conversion it needs to perform.
`converter` takes care of instantiating a graph, building the graph by parsing the JSON, and finally linking in the raw data file path if relevant (under `contentURL`).

### Tests

> The command `just test` runs all tests. The raw data used for tests is in `data/tests/` and needs to be updated if changes are made to the raw data. The validating turtle instance data is within the test scripts (in `converter/tests/`) and need to be adapted based on the changes made to the converter.

Reviewer 2: ```suggestion
The command `just test` runs all tests. The raw data used for tests is in `data/tests/` and needs to be updated if the structure of the input `json` data changes. The validating turtle instance data is within the test scripts (in `converter/tests/`) and need to be adapted based on the changes made to the converter.
```
